### ISSUE TITLE
Problem with displaying some items upon Adapter.insert() 

### ISSUE THREAD

**Issue by StriderRanger** [dhilt/ngx-ui-scroll#177]
I have a case where I need to insert multiple items in my ui-scroll list, and where each item must be treated separately to know the exact place where it must be inserted in the list.

Before inserting the items, the ui-scroll list is updated to reflect some new data change, then each item that is intended to be inserted undergoes some steps before calling `Adapter.insert()`:
- The ui-scroll list buffer is updated to contain the new item in the intended position
- The `Adapter.insert()` is called

After the insertion of each item, the `Adapter.check()` method could be called if there is some asynchronous change in the height of the item.

The problem that I am facing is that the first element to be inserted is not displayed in the DOM, as its style property is set to `position: fixed; left:-99999px`
![insert-problem](https://user-images.githubusercontent.com/15088185/82132289-c3f62600-97cd-11ea-8ed5-3f2fe86a5964.png)

Any idea why this is happening?

If it could help, here are the logs that are generated by the ui-scroll, and I added manually some lines to separate between the steps of my insert process:
[insert-problem.log](https://github.com/dhilt/ngx-ui-scroll/files/4639416/insert-problem.log)

**Note:** In the logs, you may notice the `HTML element is not defined` error after the `Adapter.check()` method. When I remove this method from the code, the display problem remains anyway, but I preferred to keep it in the logs as it might be related the non displayed item.

Please let me know if there is anything that must be clarified. 
Thanks!

**Comment by dhilt** [comment]
Let's try to figure out.

**1.** The error raises from `Item.setSize` method, which is not protected in contrast to other `Item` class methods:

https://github.com/dhilt/ngx-ui-scroll/blob/c439e1990ad03f5b3f39584700a76c0286f12fe4/src/component/classes/item.ts#L45-L59

This obviously should be fixed on the ngx-ui-scroll side.

**2.** As you have noticed, this will not help in your case. I see one interesting thing in the attached logs: calling the `Adapter.fix` method during Render process:

> stat before new items render
> adapter: fix({ updater })
> adapter: fix({ maxIndex })
> stat after new items render

Are you sure you want the fix method to be invoked at this uncertain moment?  Perhaps, it would be better to make an update (fix) when the Workflow completes its current cycle or at least its inner loop: 

```js
adapter.loopPending$.pipe(
  filter(isPending => !isPending),
  take(1)
).subscribe(() => adapter.fix({ updater }));
```

Also, updating max index manually might be not necessary, scroller updates `buffer.absMaxIndex` automatically as we may see in logs and generally this should be enough. 

**3.** The worst thing I see in logs is that "WF Cycle 2-2 STARTED" occurred twice. Is it true or maybe it is the result of manual processing you made when creating the attachment? This must not happen under any circumstances. And I'd like to have full log in this case.

**4.** The last note is about inconsistency between range and accessible items in the buffer:

> stat after new items render, pos: 7300, size: 7775.39990234375, bwd_p: 7056, fwd_p: 0, aver: 42, items: 17, range: [-7..10]

There is no doubt about the items number should be 18 after the render is done, not 17. This not-available item, as I suppose, breaks the `Adapter.check` method call, as it can't pass through the `Item.setSize` procedure I mentioned in p.1.

This lost item stays invisible in the DOM and makes the picture wrong. Render process is responsible for a) injecting new items in DOM, b) making them visible by removing `left` and `position` inline properties. But new items are already present in the Buffer before Render is started. So applying adapter-fix in non-synchronous manner might affect buffered items that are not rendered. I don't see the whole picture and how these last statements can correlate with each other, but I guess the following might help:

 - sync Adapter.fix/check calls through the technic of loopPending$ subscription
 - do not make additional min/max indexes adjustments after Adapter.insert

After this I would ask you to attach new log and let it be full (you may add lines, but not delete).

**Comment by StriderRanger** [comment]
Thank you for the clarification. As recommended I canceled the use of `Adapter.fix()` to update the max index, and I used the loopPending$ subscription to sync the `Adapter.fix()` calls.
The problem remains the same, here is the new logs without deleting anything:
[insert-problem.log](https://github.com/dhilt/ngx-ui-scroll/files/4640624/insert-problem.log)

The "WF Cycle 2-2 STARTED" occurrs twice because the `Adapter.insert()` method is called twice, and this is why:
- The 2 test elements that I intend to insert in the list are not necessarily ordered one after the other, which means that each element will take a different position in the list after that the insert process is completed. So I was obliged to handle each item separately in a loop process.

Concerning the processes that are behind the appearence of the `Adapter.fix()` method in the logs in each insert/cycle:
- The first `Adapter.fix()` that appears before the insert is used to determine the first and last loaded element in the DOM (as was recommended in [this issue](https://github.com/dhilt/ngx-ui-scroll/issues/174#))
- The second `Adapter.fix()` that appears after the insert is used to update the data that is contained in the buffer with the new ids. My goal is to keep the DOM data in sync with the buffer data, if I am doing anything wrong here or if this update is not needed at all please correct me:

```
let bufferTmp;
adapter.fix({
    updater: ({ data, $index }) => {
        bufferTmp = buffer.find(b => b[this.objectName]._id == data[this.objectName]._id);
        if (bufferTmp) {
          data.id = bufferTmp.id;
        }
        console.log("index " + $index, data)
    }
}));
```

Thank you again!

**Comment by dhilt** [comment]
I caught the issue. Two "WF Cycle 2-2 STARTED" entries was the red flag and it led me to the cause. Both `Adapter.insert` methods are invoked simply one by one and they do their job in parallel, as I see in logs: the second _insert_ run when the first one is in render phase. The second method overrun the first one. This makes the item from the 1 _insert_ rendered improperly, it remains invisible and breaks the index consistency. 

I need to think, how this can be handled on the lib end, but from the App point the fix may be just to make these 2 calls synchronous in the terms of the UiScroll Workflow cycles:

```js
adapter.insert({ /* 1 call */ })
adapter.isLoading$
  .pipe(filter(isLoading => !isLoading), take(1))
  .subscribe(() => Adapter.insert({ /* 2 call */ }));
```

The thing has came to my mind, what if the Adapter method could return a promise resolving when the Workflow cycle bound with this method invocation is done? The result code could look like

```js
await adapter.insert({ /* 1 call */ })
adapter.insert({ /* 2 call */ })
```

All the Adapter methods might behave this way.

**Comment by StriderRanger** [comment]
Making the insert/check calls synchronous solved the problem, thanks :)
Since `adapter.isLoading$` is a `Subject`, I just would like to note that it's important to test if there is no pending process first, otherwise the insert will not trigger:
```typescript
if (!adapter.isLoading) {
  Adapter.insert({ /* insert process */ })
}

adapter.isLoading$
  .pipe(filter(isLoading => !isLoading), take(1))
  .subscribe(() => Adapter.insert({ /* insert process */ }));
```
This code was based on some examples that you used before, and it comes handy in many situations.

It's a bit tricky to keep an eye on the places where the asynchronous nature of these methods may cause an unwanted behavior, so the idea you came up with is really great: Having the possibility to invoke the adapter methods with the `await` expression, by making them return a promise when the workflow cycle is done, will be of much help.

**Comment by StriderRanger** [comment]
After analyzing further the app code, using the `Adapter` methods with the `await` expression will simplify greatly the code, and make the use of these methods tremendously more simple.

Any chance to have an approximate interval when these functionality could be added to the lib?

Thank you very much for your support !

**Comment by dhilt** [comment]
I created new issue https://github.com/dhilt/ngx-ui-scroll/issues/178. My guess is 2-3 weeks might require for the release. Also, I already [pushed some code](https://github.com/dhilt/ngx-ui-scroll/pull/179) to the repository...

**Comment by dhilt** [comment]
[v1.7.0 is released](https://github.com/dhilt/ngx-ui-scroll/releases/tag/v1.7.0), it allows to protect the code after the Adapter method invocation:

```js
await adapter.insert(opt1);
adapter.insert(opt2); // safe call 2
```

Also, there is 1 new method in experimental API, which allows to protect also the first call:

```js
await adapter.relax();
await adapter.insert(opt1); // safe call 1
adapter.insert(opt2); // safe call 2
```

I didn't dare to include "relax" logic inside Adapter methods, maybe it will happen in future in some way, but I think the Adapter.relax method will persist anyway. I already started using it in the code base ([link](https://github.com/dhilt/ngx-ui-scroll/blob/92845ef057f0d99fbb9ee0a2d7246ed7f8262ba2/src/component/workflow.ts#L151)).

**Comment by StriderRanger** [comment]
That's great news. Since we have now more control on knowing when the call of the adapter methods starts or ends, it will surely help to avoid some unexpected behaviors in the app.

I will get back to you soon once I complete the implementation and test of this new feature.

Thanks!

**Comment by StriderRanger** [comment]
Implementation and tests are done, the framework works splendidly.
Thank you again for your support 👍

**Comment by bobbytechie** [comment]
@dhilt I am still getting this issue in IE Browser and issue is for both success callback and insert. IE is the main browser for us. 

Even for append also getting the same issue and only in IE Browser.

**Comment by dhilt** [comment]
@bobbytechie Hi Bobby! it would very VERY helpful if you could open new issue containing full description of the problem and minimal repro. Here is [the demo on stackblitz](https://stackblitz.com/edit/ngx-ui-scroll-1-7-0-angular-10) that you may fork.
